### PR TITLE
Document how to write a new tutorial in `CONTRIBUTING.md`

### PR DESCRIPTION
**Pull request by crcrpar**
## Motivation
We've incorporated Sphinx Gallery therefore it's a bit complicated to write a new tutorial due to the lack of description.
This pull request adds a description of (a) how the documentation is built, and (b) how to write one to CONTRIBUTING.md.

### CODE REVIEW

**Comment by crcrpar** [comment]
@nzw0301 I wonder if you could proofread this as I once explain almost the same thing to you the other day.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -108,6 +108,25 @@ extract the files in the zip to `docs/source/tutorial` directory.
...
+So if you are to add a new content to Optuna tutorial, that content should be placed in `20_recipes` and its file name should conform to the others, for example, `777_cool_feature.py`.
+The prefix numbers of files are consecutive and we expect you to follow this convention. However, this is not mandatory and if you think your content deserves the smaller number (we don’t mean anything by the order of recipes, but in general, order could convey the priority order to readers), feel free to contact Optuna committers.
+
+You may want to refer to Sphinx Gallery for the syntax of `.py` files processed by Sphinx Gallery and we follow it.
+In Optuna tutorial, there are Optuna specific conventions and limitations.
+1. 99 #s for block separation as in https://github.com/optuna/optuna/blob/2e14273cab87f13edeb9d804a43bd63c44703cb5/tutorial/10_key_features/001_first.py#L19
+2. Execution time of the new content needs to be less than three minutes. This limitation derives from Read The Docs. If your content runs some hyperparameter optimization, set the `timeout` to 180 or less.
```
Comment (nzw0301) [inline CONTRIBUTING.md:127]: Is there some link to confirm `This limitation derives from Read The Docs`?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -108,6 +108,25 @@ extract the files in the zip to `docs/source/tutorial` directory.
 
 ![image](https://user-images.githubusercontent.com/16191443/107472296-0b211400-6bb2-11eb-9203-e2c42ce499ad.png)
 
+**Writing Tutorial**  
+Tutorial is a part of Optuna’s documentation.  
+Optuna generally depends on Sphinx to build HTML files of documentation from corresponding reStructuredText (`.rst`) files in docs/source directory.
+But as you may notice, [Tutorial directory](https://github.com/optuna/optuna/tree/master/tutorial) does not have any `.rst` files, but has a bunch of Python (`.py`) files.
+We have [Sphinx Gallery](https://sphinx-gallery.github.io/stable/index.html) execute those `.py` files and generate `.rst` files with standard outputs from them and corresponding Jupyter Notebook (`.ipynb`) files.
```
Comment (nzw0301) [inline CONTRIBUTING.md:115]: ?
```suggestion
We have [Sphinx Gallery](https://sphinx-gallery.github.io/stable/index.html) that executes those `.py` files and generates `.rst` files with standard outputs from them and corresponding Jupyter Notebook (`.ipynb`) files. 
```

**Comment by nzw0301** [comment]
nit: the formatter made me confused when I wrote a tutorial https://github.com/optuna/optuna/pull/2422#issuecomment-787109344. It would be great to address it for those who write a tutorial for the first time.

File: CONTRIBUTING.md (hunk 1 above)
Comment (crcrpar) [reply CONTRIBUTING.md:127]: Can https://docs.readthedocs.io/en/stable/builds.html be?

Comment (nzw0301) [reply CONTRIBUTING.md:127]: Thank you for sharing it! I suppose so.

**Comment by crcrpar** [comment]
> nit: the formatter made me confused when I wrote a tutorial [#2422 (comment)](https://github.com/optuna/optuna/pull/2422#issuecomment-787109344). It would be great to address it for those who write a tutorial for the first time.

Thank you for reminding me of that. I understand we should make that happen but Optuna org members have been discussing how to provide formatter/static analyzer execution tool with users like https://github.com/optuna/optuna/pull/2409.

**Comment by toshihikoyanase** [comment]
@himkt Could you review this PR? If you don't have enough time, please feel free to remove the assignment.

**Comment by himkt** [comment]
@toshihikoyanase 
Sorry for the late. I'll be watching this PR but let me unassign now.
I'm not eligible to review the documentation PR due to the lack of English/Writing skills.

**Comment by toshihikoyanase** [comment]
@himkt I see. Thank you for taking your time.

@Crissman Could you proof-read the change in `CONTRIBUTING.md`?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -108,6 +108,25 @@ extract the files in the zip to `docs/source/tutorial` directory.
 
 ![image](https://user-images.githubusercontent.com/16191443/107472296-0b211400-6bb2-11eb-9203-e2c42ce499ad.png)
 
+**Writing Tutorial**
```
Comment (Crissman) [inline CONTRIBUTING.md:111]:
```suggestion
**Writing a Tutorial**  
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -108,6 +108,25 @@ extract the files in the zip to `docs/source/tutorial` directory.
 
 ![image](https://user-images.githubusercontent.com/16191443/107472296-0b211400-6bb2-11eb-9203-e2c42ce499ad.png)
 
+**Writing Tutorial**  
+Tutorial is a part of Optuna’s documentation.
```
Comment (Crissman) [inline CONTRIBUTING.md:112]:
```suggestion
Tutorials are part of Optuna’s documentation.  
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -108,6 +108,25 @@ extract the files in the zip to `docs/source/tutorial` directory.
 
 ![image](https://user-images.githubusercontent.com/16191443/107472296-0b211400-6bb2-11eb-9203-e2c42ce499ad.png)
 
+**Writing Tutorial**  
+Tutorial is a part of Optuna’s documentation.  
+Optuna generally depends on Sphinx to build HTML files of documentation from corresponding reStructuredText (`.rst`) files in docs/source directory.
```
Comment (Crissman) [inline CONTRIBUTING.md:113]:
```suggestion
Optuna depends on Sphinx to build the documentation HTML files from the corresponding reStructuredText (`.rst`) files in the docs/source directory,
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -108,6 +108,25 @@ extract the files in the zip to `docs/source/tutorial` directory.
...
+But as you may notice, [Tutorial directory](https://github.com/optuna/optuna/tree/master/tutorial) does not have any `.rst` files, but has a bunch of Python (`.py`) files.
```
Comment (Crissman) [inline CONTRIBUTING.md:114]:
```suggestion
but as you may notice, [Tutorial directory](https://github.com/optuna/optuna/tree/master/tutorial) does not have any `.rst` files. Instead, it has a bunch of Python (`.py`) files.
```

File: CONTRIBUTING.md (hunk 2 above)
Comment (Crissman) [inline CONTRIBUTING.md:115]:
```suggestion
[Sphinx Gallery](https://sphinx-gallery.github.io/stable/index.html) executes those `.py` files and generates `.rst` files with standard outputs and the corresponding Jupyter Notebook (`.ipynb`) files. 
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -108,6 +108,25 @@ extract the files in the zip to `docs/source/tutorial` directory.
...
+We have [Sphinx Gallery](https://sphinx-gallery.github.io/stable/index.html) execute those `.py` files and generate `.rst` files with standard outputs from them and corresponding Jupyter Notebook (`.ipynb`) files. 
+These generated `.rst` and `.ipynb` files are written out to docs/source/tutorial directory.
```
Comment (Crissman) [inline CONTRIBUTING.md:116]:
```suggestion
These generated `.rst` and `.ipynb` files are written to the docs/source/tutorial directory. 
```

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -108,6 +108,25 @@ extract the files in the zip to `docs/source/tutorial` directory.
...
+Tutorial is a part of Optuna’s documentation.  
+Optuna generally depends on Sphinx to build HTML files of documentation from corresponding reStructuredText (`.rst`) files in docs/source directory.
+But as you may notice, [Tutorial directory](https://github.com/optuna/optuna/tree/master/tutorial) does not have any `.rst` files, but has a bunch of Python (`.py`) files.
+We have [Sphinx Gallery](https://sphinx-gallery.github.io/stable/index.html) execute those `.py` files and generate `.rst` files with standard outputs from them and corresponding Jupyter Notebook (`.ipynb`) files. 
+These generated `.rst` and `.ipynb` files are written out to docs/source/tutorial directory. 
+The output directory (docs/source/tutorial) and source (tutorial) directory are configured in [`sphinx_gallery_conf ` of docs/source/conf.py](https://github.com/optuna/optuna/blob/2e14273cab87f13edeb9d804a43bd63c44703cb5/docs/source/conf.py#L189-L199). These generated `.rst` files are handled by Sphinx likewise the other `.rst` files. The generated `.ipynb` files are hosted on Optuna’s documentation page and downloadable (check [Optuna tutorial](https://optuna.readthedocs.io/en/stable/tutorial/index.html)).
```
Comment (Crissman) [inline CONTRIBUTING.md:117]:
```suggestion
The output directory (docs/source/tutorial) and source (tutorial) directory are configured in [`sphinx_gallery_conf ` of docs/source/conf.py](https://github.com/optuna/optuna/blob/2e14273cab87f13edeb9d804a43bd63c44703cb5/docs/source/conf.py#L189-L199). These generated `.rst` files are handled by Sphinx like the other `.rst` files. The generated `.ipynb` files are hosted on Optuna’s documentation page and downloadable (check [Optuna tutorial](https://optuna.readthedocs.io/en/stable/tutorial/index.html)).
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -108,6 +108,25 @@ extract the files in the zip to `docs/source/tutorial` directory.
...
+
+The order of contents on [tutorial top page](https://optuna.readthedocs.io/en/stable/tutorial/index.html) is determined by two keys: one is the subdirectory name of tutorial and the other is the filename (note that there are some alternatives as documented in [Sphinx Gallery - sorting](https://sphinx-gallery.github.io/stable/gen_modules/sphinx_gallery.sorting.html?highlight=filenamesortkey) but we chose this key in https://github.com/optuna/optuna/blob/2e14273cab87f13edeb9d804a43bd63c44703cb5/docs/source/conf.py#L196).
```
Comment (Crissman) [inline CONTRIBUTING.md:119]:
```suggestion
The order of contents on [tutorial top page](https://optuna.readthedocs.io/en/stable/tutorial/index.html) is determined by two keys: one is the subdirectory name of tutorial and the other is the filename (note that there are some alternatives as documented in [Sphinx Gallery - sorting](https://sphinx-gallery.github.io/stable/gen_modules/sphinx_gallery.sorting.html?highlight=filenamesortkey), but we chose this key in https://github.com/optuna/optuna/blob/2e14273cab87f13edeb9d804a43bd63c44703cb5/docs/source/conf.py#L196).  
```

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -108,6 +108,25 @@ extract the files in the zip to `docs/source/tutorial` directory.
...
+
+The order of contents on [tutorial top page](https://optuna.readthedocs.io/en/stable/tutorial/index.html) is determined by two keys: one is the subdirectory name of tutorial and the other is the filename (note that there are some alternatives as documented in [Sphinx Gallery - sorting](https://sphinx-gallery.github.io/stable/gen_modules/sphinx_gallery.sorting.html?highlight=filenamesortkey) but we chose this key in https://github.com/optuna/optuna/blob/2e14273cab87f13edeb9d804a43bd63c44703cb5/docs/source/conf.py#L196).  
+Optuna’s tutorial directory currently has two directories: (1) [10_key_features](https://github.com/optuna/optuna/tree/master/tutorial/10_key_features) which is meant to be aligned with the key features listed on [README.md](https://github.com/optuna/optuna#key-features) and explain them, and (2) [20_recipes](https://github.com/optuna/optuna/tree/master/tutorial/20_recipes) whose contents showcase how to use Optuna features conveniently.
```
Comment (Crissman) [inline CONTRIBUTING.md:120]:
```suggestion
Optuna’s tutorial directory has two directories: (1) [10_key_features](https://github.com/optuna/optuna/tree/master/tutorial/10_key_features), which is meant to be aligned with and explain the key features listed on [README.md](https://github.com/optuna/optuna#key-features) and (2) [20_recipes](https://github.com/optuna/optuna/tree/master/tutorial/20_recipes), whose contents showcase how to use Optuna features conveniently.  
```

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -108,6 +108,25 @@ extract the files in the zip to `docs/source/tutorial` directory.
...
+
+The order of contents on [tutorial top page](https://optuna.readthedocs.io/en/stable/tutorial/index.html) is determined by two keys: one is the subdirectory name of tutorial and the other is the filename (note that there are some alternatives as documented in [Sphinx Gallery - sorting](https://sphinx-gallery.github.io/stable/gen_modules/sphinx_gallery.sorting.html?highlight=filenamesortkey) but we chose this key in https://github.com/optuna/optuna/blob/2e14273cab87f13edeb9d804a43bd63c44703cb5/docs/source/conf.py#L196).  
+Optuna’s tutorial directory currently has two directories: (1) [10_key_features](https://github.com/optuna/optuna/tree/master/tutorial/10_key_features) which is meant to be aligned with the key features listed on [README.md](https://github.com/optuna/optuna#key-features) and explain them, and (2) [20_recipes](https://github.com/optuna/optuna/tree/master/tutorial/20_recipes) whose contents showcase how to use Optuna features conveniently.  
+So if you are to add a new content to Optuna tutorial, that content should be placed in `20_recipes` and its file name should conform to the others, for example, `777_cool_feature.py`.
```
Comment (Crissman) [inline CONTRIBUTING.md:121]:
```suggestion
When adding new content to the Optuna tutorials, place it in `20_recipes` and its file name should conform to the other names, for example, `777_cool_feature.py`.
```

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -108,6 +108,25 @@ extract the files in the zip to `docs/source/tutorial` directory.
...
+The prefix numbers of files are consecutive and we expect you to follow this convention. However, this is not mandatory and if you think your content deserves the smaller number (we don’t mean anything by the order of recipes, but in general, order could convey the priority order to readers), feel free to contact Optuna committers.
```
Comment (Crissman) [inline CONTRIBUTING.md:122]:
```suggestion
In general, please number the prefix for your file consecutively with the last number. However, this is not mandatory and if you think your content deserves the smaller number (the order of recipes does not have a specific meaning, but in general, order could convey the priority order to readers), feel free to propose the renumbering in your PR.
```
```suggestion
The prefix numbers of files are consecutive and we expect you to follow this convention. However, this is not mandatory and if you think your content deserves the smaller number (we don’t mean anything by the order of recipes, but in general, order could convey the priority order to readers), feel free to contact Optuna committers.
```

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -108,6 +108,25 @@ extract the files in the zip to `docs/source/tutorial` directory.
...
+
+You may want to refer to Sphinx Gallery for the syntax of `.py` files processed by Sphinx Gallery and we follow it.
```
Comment (Crissman) [inline CONTRIBUTING.md:124]:
```suggestion
You may want to refer to the Sphinx Gallery for the syntax of `.py` files processed by Sphinx Gallery.
```

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -108,6 +108,25 @@ extract the files in the zip to `docs/source/tutorial` directory.
...
+In Optuna tutorial, there are Optuna specific conventions and limitations.
```
Comment (Crissman) [inline CONTRIBUTING.md:125]:
```suggestion
Two specific conventions and limitations for Optuna tutorials:
```

**Review by Crissman — CHANGES_REQUESTED** [review]
Thanks for the tutorial guide!

**Review by Crissman — APPROVED** [review]
Thanks for the changes. LGTM.

**Comment by hvy** [comment]
@nzw0301, do you have further comments on this change? Otherwise, I think we can merge it.

**Review by nzw0301 — APPROVED** [review]

**Comment by nzw0301** [comment]
@hvy No, I don't. Thank you for ping me.

**Comment by hvy** [comment]
Got it, thanks for the quick response.

**Pull request merged**